### PR TITLE
Group invites, closes #449

### LINKED ISSUE #449: Joining a group :sparkles: a draft :sparkles:
OK, so here is the proposed workflow:
- user goes on the group page, there is only button that says "join the group"
- the user clicks the button and a message field opens with a default message like `I want to join your group as student`. the user can edit the message. There is also an explanatory text that says that the message will be send to the group for approval. And there is a "request membership" button.
- the user clicks the "request membership" button. The message field closes with a notice "Your request was sent to the group." and there is now a "membership pending" indicator instead of "join the group". 
- In the background an email is sent to the group_email and the pending membership is shown. On the group page as pending membership. duh.
- a member of the group can (un-) approve the membership. (nice to have: send an email to the new member)

### PR DESCRIPTION
**Pull request by lislis**
#### Is this solving an issue already in our backlog? If so, please mention which one
#449

#### Special things about this PR to be considered
This is one of the biggest changes yet and it will take me some time to work through this.

#### Do you have any open Questions you need to get answered?
If you find any edge cases that should be tested, please tell me!

#### Please outline the areas that should be tested, if there are any:
I'll introduce the concept of an invite. Invites are temporary relationships between a person and a group, similar to memberships but, well, temporary.
A user request membership to a group, an invite is **created**.
A group member either accepts or rejects it. Both actions **destory** the invite with the only difference being that accept also creates a membership.

So far the plan. I had some other ideas, that's why the model is a bit weird, not sure if I keep it or throw it out.

### CODE REVIEW

**Comment by lislis** [comment]
Ok, so here is a rundown of the current workflow with screenshots

- _Til Example_ logs in an on _The Rubycorns_ page clicks the `request membership` button
screenshot from 2018-08-14 23-05-53
- After clicking, there is confirmation that membership has been requested with timestamp (no screenshot)
- On _Til's_ profile page there is a list with pending invitations
screenshot from 2018-08-14 23-07-46
- _The Rubycorns_ get an email that _Til_ has requested memberhship
screenshot from 2018-08-14 23-13-06
- _Lisa Another Coach_ (who is group admin) logs into RoRganize and goes to _the Rubycorns_ group pages and sees the badge `1` at `Manage members`
screenshot from 2018-08-14 23-08-20
- On the `Manage members` page is a section with membership requests, where _Til's_ request can be accepted or rejected
screenshot from 2018-08-14 23-08-24
- Depending on whether _Til_ gets accepted or rejected they get an email and (in case of accept) become a group member
screenshot from 2018-08-14 23-10-08
screenshot from 2018-08-14 23-14-45

**Comment by lislis** [comment]
Updates! :tada:
I added feature tests, but the invites controller #destory spec is not working the way I want it to. Can someone have a :eyes: ?
Other than that I think this is ready to be reviewed! Comments welcome!


Here are the updated screens
### Request membership
screenshot from 2018-10-09 20-23-44

### Membership requested
screenshot from 2018-10-09 20-23-48

### pending memberships on profile page
screenshot from 2018-10-09 20-23-57

### New membership request when group admin
screenshot from 2018-10-09 20-25-41

### Manage requests when group admin
screenshot from 2018-10-09 20-28-03

Emails are the same

**Comment by lislis** [comment]
Ok, so now it's the fun time where tests pass locally but weird errors appear on Travis

**Comment by lislis** [comment]
So, I cleared all the caches on Travis and locally and still the same: local passing, Travis failing.
I'm out of ideas how I could reproduce the errors. @til @PragTob any ideas?

**Comment by lislis** [comment]
Yay, @vsmart figured out what caused the Travis error! Apparently I deleted a part of the vcr_cassette for groups which then caused the geocoder to fail on all tests that involved creating groups. 
Resorting that cassette from master fixed the issue! :tada:

File: app/controllers/groups_controller.rb [hunk 1]
```diff
@@ -46,6 +46,11 @@ def update
   end
 
   def show
+    @invites = Invite.where(group_id: @group.id).select do |i|
+      return false unless current_person
```
Comment (vsmart) [inline app/controllers/groups_controller.rb]: What's the diff between `signed_in` and `current_person` (which we use below)?

File: app/mailers/person_mailer.rb [hunk 2]
```diff
@@ -10,4 +10,27 @@ def new_member_email(group, person)
          subject: 'Your group has a new member!'
   end
 
+  def invitation_accepted(group, person)
+    @group = group
+    @person = person
+
+    mail to: @person.email,
+         subject: "Your membership to #{@group.name} has been accepted!"
```
Comment (vsmart) [inline app/mailers/person_mailer.rb]: spacing here looks a bit off to me, is this intended?

File: app/mailers/person_mailer.rb [hunk 3]
```diff
@@ -10,4 +10,27 @@ def new_member_email(group, person)
...
+  end
+
+  def invitation_rejected(group, person)
+    @group = group
+    @person = person
+
+    mail to: @person.email,
+         subject: "Your membership to #{@group.name} has been rejected."
```
Comment (vsmart) [inline app/mailers/person_mailer.rb]: I would rephrase this wording to `declined an invitation` instead of `reject a membership`. Same with the admin view.

File: config/environments/development.rb [hunk 4]
```diff
@@ -44,7 +44,12 @@
 
   config.action_mailer.default_url_options = { host: 'localhost:3000' }
 
+  ## If you have mailcatcher installed
+  ## (`gem install mailcatcher`, then `$ mailcatcher`),
+  ## you can use :smtp for amazeing email debugging
```
Comment (vsmart) [inline config/environments/development.rb]: typo: amazing

Comment (vsmart) [reply config/environments/development.rb]: also: should this go into the readme?

File: app/controllers/groups_controller.rb (hunk 1 above)
Comment (lislis) [reply app/controllers/groups_controller.rb]: I think I used `unless current_person` because I'm assigning the `current_user` directly on the next line. 
Otherwise there is not much difference to `signed_in?`

File: app/mailers/person_mailer.rb (hunk 2 above)
Comment (lislis) [reply app/mailers/person_mailer.rb]: Yes, aligning the arguments `to:` and `subject:`

File: app/views/groups/manage_members.html.haml [hunk 5]
```diff
@@ -1,22 +1,41 @@
...
+              = f.hidden_field :state, value: true
+              = f.submit "accept request", class: "btn btn-small btn-grey"
+            = form_for(i, html: { method: :delete }, url: invites_path(i)) do |f|
+              = f.hidden_field :state, value: false
+              = f.submit "reject request", class: "btn btn-small btn-pink"
 
-        = form_for(m, html: { method: :delete, data: { confirm: "Are you sure? Keep in mind, destruction is irreversible!" } }, url: membership_path(m)) do |f|
-          = f.submit "remove person from group", class: "btn btn-small btn-pink"
+    - else
+      %h2.page-header 📪 Membership requests
```
Comment (sareg0) [inline app/views/groups/manage_members.html.haml]: Maybe we could move the `%h2.page-header 📪 Membership requests` outside of the if/else, since it is used the same in both conditions

File: app/mailers/person_mailer.rb (hunk 3 above)
Comment (lislis) [reply app/mailers/person_mailer.rb]: Good point :+1:

File: config/environments/development.rb (hunk 4 above)
Comment (lislis) [reply config/environments/development.rb]: I'll add a note to the readme

File: app/views/groups/manage_members.html.haml (hunk 5 above)
Comment (lislis) [reply app/views/groups/manage_members.html.haml]: We talked about it, because it's actually two different emoji for the two different cases ;)

File: app/views/person_mailer/new_invitation.html.haml [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+%p
+  = @person.full_name
+  has requested to join your group!
+
+%p
+  You can find their profile here:
+  \#{link_to "profile page", person_url(@person)}.
+
+%p
+  You can either accept oder reject the request
```
Comment (vsmart) [inline app/views/person_mailer/new_invitation.html.haml]: ODER REJECT

Comment (lislis) [reply app/views/person_mailer/new_invitation.html.haml]: ODER WAS

File: app/views/groups/manage_members.html.haml (hunk 5 above)
Comment (sareg0) [reply app/views/groups/manage_members.html.haml]: aha!
